### PR TITLE
Add {{map-by}} and {{filter-by}} helpers

### PR DESCRIPTION
**Pull request by martndemus**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,49 @@ Pipes the return values of actions in a sequence of actions.
...
+### Sort By
+Sort an array by given properties.
+
+```hbs
+{{#each (sort-by users 'lastName' 'firstName') as |user|}}
+  {{user.lastName}}, {{user.firstName}}
+{{/each}}
+```
+
+You can append `:desc` to properties to sort in reverse order.
```
Comment (poteto) [inline README.md]: Add example for this?

File: README.md [hunk 2]
```diff
@@ -26,6 +26,49 @@ Pipes the return values of actions in a sequence of actions.
 </button>
 ```
 
+### Map By
+Maps an array on a property.
+
+```hbs
+{{#each (map-by users 'fullName') as |name|}}
```
Comment (poteto) [inline README.md]: Small nitpick but we prefer double quotes in his!

File: addon/helpers/filter-by.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+      } else {
+        filterFn = (item) => isPresent(get(item, byPath));
+      }
+
+      let cp = filter(`array.@each.${byPath}`, filterFn);
+
+      defineProperty(this, 'content', cp);
+    } else {
+      defineProperty(this, 'content', null);
```
Comment (poteto) [inline addon/helpers/filter-by.js]: Would this make sense to return an empty array here?

File: addon/helpers/filter-by.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    }
+  }),
+
+  contentDidChange: observer('content', function() {
```
Comment (poteto) [inline addon/helpers/filter-by.js]: `content.[]`?

Comment (martndemus) [inline addon/helpers/filter-by.js]: Probably yeah.

File: addon/helpers/filter-by.js (hunk 3 above)
Comment (martndemus) [inline addon/helpers/filter-by.js]: Yes.

File: addon/helpers/filter-by.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+export default Helper.extend({
+  compute([array, byPath, value]) {
+    set(this, 'array', array);
+    set(this, 'byPath', byPath);
+    set(this, 'value', value);
+
+    return get(this, 'content');
+  },
+
+  array:   null,
```
Comment (poteto) [inline addon/helpers/filter-by.js]: Small nitpick: prefer not to align like this, makes adding/removing keys a bit annoying

File: addon/helpers/filter-by.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        }
+      } else {
+        filterFn = (item) => isPresent(get(item, byPath));
+      }
+
+      let cp = filter(`array.@each.${byPath}`, filterFn);
+
+      defineProperty(this, 'content', cp);
+    } else {
+      defineProperty(this, 'content', []);
```
Comment (poteto) [inline addon/helpers/filter-by.js]: How about we move this to the top to return early, so we don't need an `if/else` and just an `if`:

```js
if (isEmpty(byPath)) {
  defineProperty(this, 'content', []);
  return;
}

let filterFn;

if (value) {
  ...
}
```

File: addon/helpers/map-by.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+export default Helper.extend({
+  compute([array, byPath, value]) {
+    set(this, 'array', array);
+    set(this, 'byPath', byPath);
+
+    return get(this, 'content');
+  },
+
+  array:   null,
```
Comment (poteto) [inline addon/helpers/map-by.js]: Prefer not to align like this

File: addon/helpers/map-by.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  content: null,
+
+  byPathDidChange: observer('byPath', function() {
+    let byPath = get(this, 'byPath');
+
+    if (byPath) {
+      let cp = mapBy('array', byPath);
+      defineProperty(this, 'content', cp);
+    } else {
+      defineProperty(this, 'content', []);
```
Comment (poteto) [inline addon/helpers/map-by.js]: Maybe early return?

File: addon/helpers/sort-by.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    set(this, 'array', array);
+    set(this, 'sortProps', sortProps);
+
+    return get(this, 'content');
+  },
+
+  array: null,
+  sortProps: null,
+
+  sortPropsDidChange: observer('sortProps', function() {
```
Comment (poteto) [inline addon/helpers/sort-by.js]: I think this might be an array, so maybe `sortProps.[]`?

File: addon/helpers/sort-by.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    let sortProps = get(this, 'sortProps');
+
+    if (isArray(sortProps) && isPresent(sortProps)) {
+      defineProperty(this, 'content', sort('array', 'sortProps'));
+    } else {
+      defineProperty(this, 'content', []);
```
Comment (poteto) [inline addon/helpers/sort-by.js]: Maybe early return?

File: addon/helpers/sort-by.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    }
+  }),
+
+  contentDidChange: observer('content', function() {
```
Comment (poteto) [inline addon/helpers/sort-by.js]: `content.[]`?

File: addon/helpers/map-by.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    }
+  }),
+
+  contentDidChange: observer('content', function() {
```
Comment (poteto) [inline addon/helpers/map-by.js]: `content.[]`?

File: addon/helpers/filter-by.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+  }),
+
+  contentDidChange: observer('content', function() {
```
Comment (poteto) [inline addon/helpers/filter-by.js]: `content.[]`?

File: tests/integration/helpers/filter-by-test.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+const { A: emberArray, run, set } = Ember;
+
+moduleForComponent('filter-by', 'Integration | Helper | {{filter-by}}', {
+  integration: true
+});
+
+test('It filters by value', function(assert) {
+  this.set('array', emberArray([
+    { foo: true,  name: 'a' },
```
Comment (poteto) [inline tests/integration/helpers/filter-by-test.js]: Prefer not to align like this

File: tests/integration/helpers/filter-by-test.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    { foo: true,  name: 'c' }
+  ]));
+
+  this.render(hbs`
+    {{~#each (filter-by array 'foo' true) as |item|~}}
+      {{~item.name~}}
+    {{~/each~}}
+  `);
+
+  assert.equal(this.$().text().trim(), 'ac');
```
Comment (poteto) [inline tests/integration/helpers/filter-by-test.js]: Assertion messages please

File: tests/integration/helpers/map-by-test.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    { name: 'c' }
+  ]));
+
+  this.render(hbs`
+    {{~#each (map-by array 'name') as |name|~}}
+      {{~name~}}
+    {{~/each~}}
+  `);
+
+  assert.equal(this.$().text().trim(), 'abc');
```
Comment (poteto) [inline tests/integration/helpers/map-by-test.js]: Assertion messages please

File: tests/integration/helpers/sort-by-test.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    { name: 'b' }
+  ]));
+
+  this.render(hbs`
+    {{~#each (sort-by array 'name') as |user|~}}
+      {{~user.name~}}
+    {{~/each~}}
+  `);
+
+  assert.equal(this.$().text().trim(), 'abc');
```
Comment (poteto) [inline tests/integration/helpers/sort-by-test.js]: Assertion messages please!

**Comment by poteto** [comment]
LGTM, mostly minor nitpicks!

File: app/helpers/filter-by.js [hunk 18]
```diff
@@ -0,0 +1 @@
+export { default, filterBy } from 'ember-functional-helpers/helpers/filter-by';
```
Comment (poteto) [inline app/helpers/filter-by.js]: Can we also export in https://github.com/DockYard/ember-functional-helpers/blob/remove-compose/addon/index.js please

**Comment by martndemus** [comment]
@homu r+

**Pull request merged**